### ISSUE TITLE
Cache and report on git version incompatibilities

### ISSUE THREAD

**Issue by phil-s** [magit/magit#4065]
Commit f6307724251b1632efb7607d187dd8ccc602ae55 added a repository-local cache of the availability of one specific feature, based on the version of git being used.

It's long been the case that trying to use Magit with an unsupported older version of git causes problems which can be difficult for users to interpret.  Most recently this cropped up in https://github.com/magit/magit/issues/3714#issuecomment-596500488 and so I'm assuming that this kind of problem is still a going concern.

I suggest adapting that repository-local caching approach so that any and all version-related issues can be captured (as a one-time action), and reported on.

If the cache doesn't exist yet, we would check the git version, then cache store all related information.  To start with this would be (a) the git version; (b) whether that version is supported by Magit; and (c) the diff-ita-kludge-p value from the aforementioned commit.

If the version is unsupported then Magit can make a loud complaint to the user pointing out the incompatibility in an unambiguous fashion (and also pointing out the 'zap caches' command, in case the user wishes to forcibly re-check, perhaps after upgrading git).

**Comment by tarsius** [comment]
Sounds like a good idea. Could you implement that please?

**Comment by kyleam** [comment]
> It's long been the case that trying to use Magit with an unsupported older version of git causes problems which can be difficult for users to interpret. Most recently this cropped up in [#3714 (comment)](https://github.com/magit/magit/issues/3714#issuecomment-596500488) and so I'm assuming that this kind of problem is still a going concern.

Hmm, `magit-startup-asserts` should display an informative message when the git version on the local machine isn't new enough.  (I quickly tested by tweaking `magit--minimal-git` to the currently unborn 2.27.0, and I saw the warning.)  At first I thought that perhaps a warning wasn't occurring because the user was operating over Tramp, but then looking closer I think `magit--tramp-asserts` should catch an unsupported git in that case.

> To start with this would be [...]

There are a few other spots that could benefit from this.  There's a cached `update-index-has-ignore-sw-p` as of b857fc1b.  And looking through `magit-git-version` calls, other candidates for cached git version conditions are `magit-fetch-modules`, `magit-rev-name` `magit-update-ref`, `magit--tramp-asserts`, `magit-submodule-add-1`, and `magit-submodule-remove`.

**Comment by phil-s** [comment]
I've had another brief look at this.  Ideally there would be a very small (yet dependable) set of places where we would need to perform the "have we done these tests yet?" check, and it wasn't immediately clear to me what those would be.  We could expect that *most* of the time people are going to start out with `magit-status`, so merely acting there would be fairly effective; but as a user's initial magit interaction with some repository *might* be any arbitrary magit command, it would be good to catch "the first attempt to call git" for any given repository, regardless of what caused it.

For the calls to git for side-effects, I think `magit-process-setup` would be the right place?

For the others it's not quite as clear, and even the `magit-git-debug` code (which has a similar use-case of observing 'all' calls to git) is explicitly ignored in some cases.

Do all such calls use `(apply #'magit-process-file magit-git-executable ...)`?

```
./magit-git.el:335:  (apply #'magit-process-file magit-git-executable nil nil nil
./magit-git.el:356:      (and (zerop (apply #'magit-process-file magit-git-executable nil t nil
./magit-git.el:377:             (status (apply #'magit-process-file magit-git-executable
./magit-git.el:404:      (apply #'magit-process-file magit-git-executable nil (list t nil) nil
./magit-git.el:415:      (apply #'magit-process-file magit-git-executable nil (list t nil) nil
./magit-git.el:450:              (let ((exit (apply #'magit-process-file magit-git-executable
./magit-git.el:469:    (apply #'magit-process-file magit-git-executable
./magit-extras.el:86:    (apply #'magit-process-file magit-git-executable nil 0 nil "gui" "blame"
```

If that's everything, I'm wondering whether (a) `magit-process-file` itself is a reasonable place to act (but we'd need to test whether it was going to call `git`); or (b) if there should be another wrapper to perform the `magit-process-file` call, but also act as a central place to put "we're about to call git" functionality.

**Comment by tarsius** [comment]
That sounds reasonable but I haven't looked at this in detail yet.

Is there some work that could be done now without deciding on where to do the check that wold not be at risk of becoming useless if the details change?

**Comment by phil-s** [comment]
Yes, I think most of the code can be written without needing to know those details, so I could make a start on that (next week at the earliest).  Existing code which would explicitly check for a particular Git feature can trigger the information gathering phase if it hasn't already occurred, so I think we can adapt existing use-cases without worrying about the more general case of catching any possible attempt to use a too-old Git version.

**Comment by phil-s** [comment]
I've started on this.

Can anyone confirm for me whether the reason that `magit-submodule-remove` requires git 2.12.0 is purely the use of `(magit-git "submodule" "absorbgitdirs" "--" modules)` -- a version dependency which is more specifically flagged in `magit-submodule-add-1` -- or if there are additional calls in `magit-submodule-remove` which are also a factor?

If it's the same thing in both cases, then I can reduce it to a single 'feature' to check for.

This is WIP, but the initial code is along these lines:

```el
(defun magit-git-feature-list ()
  "Return the features supported by this repository's git version."
  (when-let ((git-version (magit-git-version)))
    (let (feature-list)
      ;; The minimum Emacs version supported by Magit.
      (push (cons 'minimal-emacs (version<= magit--minimal-emacs emacs-version))
            feature-list)
      ;; The minimum Git version for various features.
      (dolist (item `((,magit--minimal-git . minimal-git)
                      ("2.8.0" . magit-fetch-modules-jobs)
                      ("2.12.0" . magit-submodule-remove) ;; possibly the same as `submodule-absorbgitdirs' ?
                      ("2.12.0" . submodule-absorbgitdirs) ;; as ^it^ calls submodule absorbgitdirs itself.
                      ("2.13.0" . name-rev-exclude)
                      ("2.19.0" . diff-ita-kludge)
                      ("2.25.0" . update-index-has-ignore-sw)
                      ))
        (cl-destructuring-bind (minimum . feature) item
          (push (cons feature (version<= minimum git-version))
                feature-list)))
      ;; Cache and return the feature list.
      (magit-repository-local-set 'git-features feature-list)
      feature-list)))

(defun magit-git-feature-supported-p (feature)
  "Non-nil when the current repository's git version supports FEATURE."
  (let* ((cache (magit-repository-local-exists-p 'git-features))
         (git-features (if cache
                           (cdr cache)
                         (magit-git-feature-list))))
    (alist-get feature git-features)))
```

Looking at it now, I suspect I'll move all the cache management into `magit-git-feature-list`, but that's just details.  The names are essentially all placeholders too (and I'm wondering whether something as generic as `magit-can-use-p` would be a better predicate name).

The intent is then to make changes like:

```diff
--- lisp/magit-submodule.el
+++ #<buffer magit-submodule.el>
@@ -207,7 +207,7 @@
              (magit-process-sentinel process event)
            (process-put process 'inhibit-refresh t)
            (magit-process-sentinel process event)
-           (unless (version< (magit-git-version) "2.12.0")
+           (when (magit-git-feature-supported-p 'submodule-absorbgitdirs)
              (magit-call-git "submodule" "absorbgitdirs" path))
            (magit-refresh)))))))
```

**Comment by kyleam** [comment]
> Can anyone confirm for me whether the reason that `magit-submodule-remove` requires git 2.12.0 is purely the use of `(magit-git "submodule" "absorbgitdirs" "--" modules)`

Quickly looking through `magit-submodule-remove`, I don't spot anything else that came with 2.12.0.  (It looks like there is an additional incompatibility though: `magit-submodule-remove` makes a call to `git stash push`, which wasn't introduced until 2.13.0).

**Comment by phil-s** [comment]
Thanks Kyle.

Current WIP is https://github.com/magit/magit/compare/fd6466025725f8780b7da18639ded1b7b141de26...phil-s:phil-s/git-version-compatibility

**Comment by tarsius** [comment]
Ping!

Do you still plan to get back to this?

**Comment by phil-s** [comment]
I do.  I was reminding myself a couple of days ago, in fact.

IIRC the current code 'works' but is too aggressive in practice, and so I had work to do to stop it from caring quite so much about compatibility all of the time at such a low level.

**Comment by tarsius** [comment]
(Have fun with the rebasing :grin: )

**Comment by tarsius** [comment]
Related: #4578.

**Comment by phil-s** [comment]
I did have another play with this recently.  My initial implementation had proved to be too aggressive, such that I was triggering errors before the pre-existing start-up warnings had a chance to be displayed.  To work around that, I've recently added a small number of special-cases to inhibit my checks for certain calls, and that seemed to be working in my initial tests.  I've pushed a WIP commit of those changes to https://github.com/phil-s/magit/commits/phil-s/git-version-compatibility

https://github.com/phil-s/magit/commit/a55b5393ca5159f6da3127f9a7d5085f0ca649f1 shows the main approach to testing features.

**Comment by tarsius** [comment]
Did you see that I just opened #4603?

If you still think we should go further than that, then please first rebase your branch onto `master`. Then also look into whether it might make sense to rebase on top of my proposed changes.

**Comment by tarsius** [comment]
I've just merged my changes as I think they go far enough.

If you would still like to go further then you would have to rebase on top of those changes now.

**Comment by phil-s** [comment]
All good.  Hmm.  I was sure I'd replied to your previous comment, but apparently not.  It had only been to say that I had indeed noticed the new PR, and that was why I'd pushed that WIP commit, in case you'd wanted to take a look.  I haven't had an opportunity to check out your new code, but I'll try to do so.

**Comment by tarsius** [comment]
I am closing this for now, but if it turns out that there's more work to be done, then we can of course reopen.
